### ISSUE TITLE
(Unnamed) semaphore issues when  using ISIO through pybind.

### ISSUE THREAD

**Issue by DasVinch** [milk-org/ImageStreamIO#50]
Hey guys - in particular @a-sevin and @drbitboy .

TLDR: see other comments below. It's 4-byte alignement at work.

Since merging Brian's PR #40 and transitioning to embedded semaphores, I'm having a misbehavior when calling SHMs through pyMilk/ImageStreamIOWrap.
Calling a semwait upon a semaphore that's not immediately passing (i.e., sem == 0) causes a futex crash.
Unclear to me what the heck is going on... here's a minimum crashing example.

```python
import numpy as np
from ImageStreamIOWrap import Image
print('Create')
img = Image()
img.create('something', np.array([0], np.uint16), location=-1, shared=1, NBkw=50)
print('OK.')

print('semwait get.')
for k in range(10):
    print(k)
    img.semtimedwait(0, 0.1)
print('OK.')

----
Create
OK.
semwait get.
0
1
The futex facility returned an unexpected error code.
Aborted (core dumped)
```

However... the same in straight C isn't a problem...

```C
#include "ImageStreamIO.h"
#include "ImageStruct.h"

#include <time.h>


int main() {
    char name[] = "asdf_ghjjk";

    IMAGE image;

    uint32_t dims[3] = {2,3,4};

    ImageStreamIO_createIm(&image, name, 2, dims, _DATATYPE_INT16, 1, 50, 0);
    ImageStreamIO_destroyIm(&image);
    ImageStreamIO_createIm(&image, name, 2, dims, _DATATYPE_INT16, 1, 50, 0);
    ImageStreamIO_closeIm(&image);
    ImageStreamIO_openIm(&image, name);


    int sem_idx = ImageStreamIO_getsemwaitindex(&image, 0);

    struct timespec t;

    float timeout_sec = 0.1;

    for (int i = 0; i < 20; i++) {
        printf("sem_timedwait %d...", i);
        clock_gettime(CLOCK_ISIO, &t);
        t.tv_nsec += (long)(timeout_sec * 1000000000L);
        t.tv_sec += t.tv_nsec / 1000000000L;
        t.tv_nsec = t.tv_nsec % 1000000000L;   
        ImageStreamIO_semtimedwait(&image, sem_idx, &t);
        printf("ok.\n");
    }

    return 0;
}
--------
$ ./isio_test 
sem_timedwait 0...ok.
sem_timedwait 1...ok.
sem_timedwait 2...ok.
sem_timedwait 3...ok.
sem_timedwait 4...ok.
sem_timedwait 5...ok.
sem_timedwait 6...ok.
sem_timedwait 7...ok.
sem_timedwait 8...ok.
sem_timedwait 9...ok.
sem_timedwait 10...ok.
sem_timedwait 11...ok.
sem_timedwait 12...ok.
sem_timedwait 13...ok.
sem_timedwait 14...ok.
sem_timedwait 15...ok.
sem_timedwait 16...ok.
sem_timedwait 17...ok.
sem_timedwait 18...ok.
sem_timedwait 19...ok.
```

Any thoughts....?

Unfortunately, this breaks pyMlk, which is fairly mission-critical in a number of spots.
The most obvious is a variety of mismatches between duplicate compiled libraries... I hope I have been careful enough...?

Vince

**Comment by DasVinch** [comment]
Update:

It works with 32 bit datatypes... only broken with np.uit8, np.int8, np.int16, uint16.
I smell a data alignment issue...

The plot thickens.

edit: if the datatype is uint16 but the size is 2 (and not 1 as above), it works.
Data alignment indeed?

**Comment by DasVinch** [comment]
Confirmed:
With `uint32_t dims[3] = {3,3};` for an odd size in the C example, it performs a beautiful futex crash as well.

I'm pretty sure this is due to an x4 byte data alignment problem. Possibly in the new size computation methods? Possibly we had it before, but we never noticed because the memory overstepping was in some section of the metadata we never bothered to care about. But now it seems to mess up the newly embedded semaphores!!

**Comment by a-sevin** [comment]
Ok... It's inside the C...

changing in your sample : 

```C
    uint32_t dims[1] = {1};
    ImageStreamIO_createIm(&image, name, 1, dims, _DATATYPE_INT16, 1, 50, 0);
```

It's locked in:
```
__futex_abstimed_wait_common64 (private=<optimized out>, cancel=true, abstime=0x7fffffffd1a0, op=265, expected=0, futex_word=0x7ffff7fbaa82) at ./nptl/futex-internal.c:57
```

**Comment by drbitboy** [comment]
For the memory mapping, I am pretty sure I could create a macro or function that increments the (uint8*)map pointer to ensure 8-byte boundaries for each data member.

Another option would be to re-order the metadata items by decreasing element size (8-byte items, 4-byte, 2-byte, 1-byte (strings). 

I was wondering when this would show up, which was one reason I moved the metadata address calculations to a single routine, i.e. so we would only have to fix this in one place instead of three.
